src/netlink.c

```diff
@@ -0,0 +1,385 @@
+static int meth_send(lua_State *L) {
+    p_netlink nl = (p_netlink)auxiliar_checkclass(L, "netlink{connected}", 1);
+    size_t payload_size;
+    const char *payload = luaL_checklstring(L, 2, &payload_size);
+    int flags = luaL_optinteger(L, 3, 0);
+    struct nlmsghdr *nlh = (struct nlmsghdr *)malloc(NLMSG_SPACE(payload_size));
+    p_timeout tm = &nl->tm; 
+    size_t sent = 0;
+    int err;
+
+    if (nlh == NULL) {
+        lua_pushnil(L);
+        lua_pushliteral(L, "error allocating memory in send()");
+        return 2;
+    }
+ 
+    memset(nlh, 0, NLMSG_SPACE(payload_size));
+    nlh->nlmsg_len = NLMSG_SPACE(payload_size);
+    nlh->nlmsg_pid = nl->srcpid;
+    nlh->nlmsg_flags = flags;
+    memcpy(NLMSG_DATA(nlh), payload, payload_size);
+
+    err = socket_send(&nl->fd, (char *)nlh, nlh->nlmsg_len, &sent, tm);
+    if (err != IO_DONE) {
+        lua_pushnil(L);
+        lua_pushliteral(L, "error sending message");
+        return 2;
+    }
+
+    lua_pushinteger(L, sent);
+    return 1;
+}
+
+/*-------------------------------------------------------------------------*\
+* Send data through unconnected netlink socket
+\*-------------------------------------------------------------------------*/
+static int meth_sendto(lua_State *L) {
+    p_netlink nl = (p_netlink)auxiliar_checkclass(L, "netlink{unconnected}", 1);
+    size_t payload_size;
+    const char *payload = luaL_checklstring(L, 2, &payload_size);
+    int dstpid = luaL_checkinteger(L, 3);
+    int groups = luaL_optinteger(L, 4, 0);
+    int flags = luaL_optinteger(L, 5, 0);
+    struct nlmsghdr *nlh = (struct nlmsghdr *)malloc(NLMSG_SPACE(payload_size));
+    struct sockaddr_nl addr;
+    p_timeout tm = &nl->tm; 
+    size_t sent = 0;
+    int err;
+
+    if (nlh == NULL) {
+        lua_pushnil(L);
+        lua_pushliteral(L, "error allocating memory in send()");
+        return 2;
+    }
+
+    memset(&addr, 0, sizeof(addr));
+    addr.nl_pid = dstpid;
+    addr.nl_family = AF_NETLINK;
+    addr.nl_groups = groups;
+
+    memset(nlh, 0, NLMSG_SPACE(payload_size));
+    nlh->nlmsg_len = NLMSG_SPACE(payload_size);
+    nlh->nlmsg_pid = nl->srcpid;
+    nlh->nlmsg_flags = flags;
+    memcpy(NLMSG_DATA(nlh), payload, payload_size);
+
+    err = socket_sendto(&nl->fd, (char *)nlh, nlh->nlmsg_len, &sent, 
+            (SA *)&addr, sizeof(addr), tm);
+    if (err != IO_DONE) {
+        lua_pushnil(L);
+        lua_pushliteral(L, "error sending message");
+        return 2;
+    }
+
+    lua_pushinteger(L, sent);
+    return 1;
+}
+
+/*-------------------------------------------------------------------------*\
+* Receives data from a netlink socket
+\*-------------------------------------------------------------------------*/
+static int meth_receive(lua_State *L) {
+    p_netlink nl = (p_netlink)auxiliar_checkclass(L, "netlink{connected}", 1);
+    struct nlmsghdr *nlh = (struct nlmsghdr *)malloc(NLMSG_SPACE(MAX_PAYLOAD));
```

Comment:
why using malloc?

Reply:
I need to allocate a sufficiently large buffer to store the message. Do you suggest using something else?